libopenarc/arc.c

```diff
@@ -3560,7 +3560,74 @@ arc_get_domain(ARC_MESSAGE *msg)
 */
 
 const char *
-arc_chain_str(ARC_MESSAGE *msg)
+arc_chain_status_str(ARC_MESSAGE *msg)
 {
 	return arc_code_to_name(chainstatus, msg->arc_cstate);
 }
+
+/*
+**	ARC_CHAIN_CUSTODY_STR -- retrieve domain chain, as a string
+**
+**	Parameters:
+**      msg -- ARC_MESSAGE object
+**      buf -- where to write
+**      buflen -- bytes at "buf"
+**
+**	Return value:
+**	    Number of bytes written. If value is greater than or equal to buflen
+**		argument, then buffer was too small and output was truncated.
+*/
+int
+arc_chain_custody_str(ARC_MESSAGE *msg, u_char *buf, size_t buflen)
+{
+	u_int set;
+	u_char *instance;
+	ARC_KVSET *kvset;
+	char *str = NULL;
+	struct arc_dstring *tmpbuf;
+	int appendlen = 0;
+
+	tmpbuf = arc_dstring_new(msg, BUFRSZ, MAXBUFRSZ);
+	if (tmpbuf == NULL)
+	{
+		arc_dstring_free(tmpbuf);
+		arc_error(msg, "failed to allocate dynamic string");
+		return ARC_STAT_NORESOURCE;
+	}
+
+	assert(msg != NULL);
+	assert(buf != NULL);
+	assert(buflen > 0);
+
+	memset(buf, '\0', buflen);
+
+	for (set = msg->arc_nsets; set > 0; set--)
+	{
+		for (kvset = arc_set_first(msg, ARC_KVSETTYPE_SEAL);
+		     kvset != NULL;
+		     kvset = arc_set_next(kvset, ARC_KVSETTYPE_SEAL))
+		{
+			instance = arc_param_get(kvset, "i");
+			if (atoi(instance) == set)
+				break;
+		}
+
+		str = arc_param_get(kvset, "d");
+		if (str == NULL)
+		{
+			continue;
+		}
+
+		if (set < msg->arc_nsets)
+		{
+			(void) arc_dstring_printf(tmpbuf, ":%s", str);
```

Comment:
This could be simplified to one line:

(void) arc_dstring_printf(tmpbuf, "%s%s", (set < msg->arc_nsets ? ":" : "", str));